src/Microsoft.FeatureManagement/FeatureManager.cs

```diff
@@ -241,77 +241,108 @@ public ValueTask<Variant> GetVariantAsync(string feature, TargetingContext conte
         {
-            bool isFeatureEnabled = false;
+            EvaluationEvent evaluationEvent = new EvaluationEvent
+            {
+                FeatureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false)
+            };
 
-            FeatureDefinition featureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false);
+            await EvaluateFeature(evaluationEvent, context, useContext, cancellationToken);
 
-            VariantDefinition variantDefinition = null;
+            return evaluationEvent.Variant;
+        }
 
-            if (featureDefinition != null)
+        private async ValueTask<bool> IsEnabledEvaluation<TContext>(string feature, TContext appContext, bool useAppContext, CancellationToken cancellationToken)
+        {
+            EvaluationEvent evaluationEvent = new EvaluationEvent
             {
-                isFeatureEnabled = await IsEnabledAsync(featureDefinition, appContext, useAppContext, cancellationToken).ConfigureAwait(false);
+                FeatureDefinition = await GetFeatureDefinition(feature).ConfigureAwait(false)
+            };
+
+            await EvaluateFeature(evaluationEvent, appContext, useAppContext, cancellationToken);
+
+            return evaluationEvent.IsEnabled;
+        }
+
+        private async Task<EvaluationEvent> EvaluateFeature<TContext>(EvaluationEvent evaluationEvent, TContext context, bool useContext, CancellationToken cancellationToken)
+        {
+            if (evaluationEvent.FeatureDefinition != null)
+            {
+                //
+                // Determine IsEnabled
+                evaluationEvent.IsEnabled = await IsEnabledAsync(evaluationEvent.FeatureDefinition, context, useContext, cancellationToken).ConfigureAwait(false);
 
-                if (featureDefinition.Variants != null && featureDefinition.Variants.Any() && featureDefinition.Allocation != null)
+                //
+                // Determine Variant
+                VariantDefinition variantDefinition;
+
+                if (evaluationEvent.FeatureDefinition.Allocation == null || (!evaluationEvent.FeatureDefinition.Variants?.Any() ?? false))
                 {
-                    if (!isFeatureEnabled)
+                    variantDefinition = null;
+
+                    evaluationEvent.AssignmentReason = AssignmentReason.None;
+                }
+                else
+                {
+                    if (!evaluationEvent.IsEnabled)
                     {
-                        variantDefinition = featureDefinition.Variants.FirstOrDefault((variant) => variant.Name == featureDefinition.Allocation.DefaultWhenDisabled);
+                        variantDefinition = evaluationEvent.FeatureDefinition.Variants.FirstOrDefault((variant) => variant.Name == evaluationEvent.FeatureDefinition.Allocation.DefaultWhenDisabled);
+
+                        evaluationEvent.AssignmentReason = AssignmentReason.DisabledDefault;
```

Comment:
If there is no default variant for the disabled feature, should the reason be None or DisabledDefault?